Replaced by PR#4 with cleaned commits

gingerlavender opened the pull request:
Added two following interfaces: 
- ILottery, which is a list of events and functions for interaction with end users;
- ILotteryErrors, which contains all the errors related to the lottery contract,

along with Types library containing LotteryStatus enum.

At any point of time lottery can be in one of the five statuses (or states), but important to notice that status itself is not set manually, instead it is derived from actual state variables during call time (which helps, for example, to switch from OpenedForRegistration to RegistrationEnded without any manual interaction).
These are:
- Closed (no activity at all)
- OpenedForRegistration (lottery has been started and registration time (which may be extended) has not ended yet, so users can enter and quit the lottery at this (and only this) phase)
- RegistrationEnded (registration is closed, but the request to oracle to determine a winner has not been sent yet)
- WaitingForReveal (the request has been sent and we are waiting oracle to call our fulfillRandomWords() function)
- Invalid (current logic is to set it if registration time has ended, and participants number is less than target amount, but later it can also be used to reflect oracle timeouts and other force majeure circumstances)

If current status is Invalid, privileged account can close that lottery, and participants will be able to get refund for their ticket any time (but only until deadline expiration, later that funds can be collected by the person with such access).

If fulfillRandomWords() has been successfully called, then winner is revealed, NFT is minted, money from ticket sales can be withdrawn by someone with corresponding permissions, lottery state is reset, and status is Closed again.

It might be worth considering what happens if mint in callback will not be successful (e.g. winner is a smart contract without onERC721Received() implementation): the whole lottery invalidation seems too harsh, so it probably makes sense to mint NFT to safe address in this case or select a couple of fallback winners. 

Speaking about contract management, plain Ownable is likely not enough and AccessManager sounds like overkill, so AccessControl might be quite suitable.

Simple lottery state diagram can be found below:

<img width="2250" height="970" alt="Lottery state diagram" src="https://github.com/user-attachments/assets/52665c29-fd80-449d-93e6-2d1efd3b686f" />

gingerlavender committed:
Changed solc version to 0.8.30

gingerlavender committed:
Lottery contract (lottery proccess flow, users interaction, funds withdraw by organizer) and errors interfaces, Types library containing LotteryStatus

gingerlavender committed:
clarity in quit()

gingerlavender committed:
changed ExpiredRefundCollected arguments considering ticket price constant for now

gingerlavender committed:
Fixed LotteryStatus documentation

bodrovis commented:
Good job! It would be nice to explain how to run this locally (e.g. list of commands). Specifically, our front-end team will require a local blockchain with the contracts deployed

gingerlavender committed:
Mocks and instructions to set up contracts locally

gingerlavender committed:
Update README

gingerlavender committed:
Fixed subscription funding issues, edited local setup instructions list

gingerlavender commented:
Added possibility (and instructions how) to run this in a local blockchain, using GuideDAO token and VRF coordinator mocks to simulate interactions in the future. However, LotteryDeployScript and VRF consumer config could be useful for production/public testnets deployments as well

gingerlavender committed:
Test contract draft

gingerlavender committed:
Added LotteryConfig

gingerlavender committed:
Updated interfaces:

- Added lotteryNumber to events depending on it
- Added mutability for organizer and price (setTicketPrice() and setOrganizer()) and corresponding events and errors

gingerlavender committed:
Updated lottery deployment script:

- Removed NUM_WORDS from deployment configuration, as it is always equal to one
- Passing initial organizer and ticket price to Lottery constructor

gingerlavender committed:
Added a few words about LotteryConfig

gingerlavender committed:
Started implementing contract functionality.

* Added basic structures:
- ParticipantInfo
- RefundBatch
- LotteryState

* Made draft for some constants with dummy values (business logic related variables, old lottery data cleanup limitations)

* Organized the way participants are stored (participants, participantsInfo)
* Organized the way refund batches are stored and tied to former participants

* Implemented most of the 'lottery flow operator' functions
* Implemented enter() function
* Implemented refund logic
* Implemented lottery status derivation
* Implemented possibility to clear old data about former lottery participants (for lotteries that were held a defined constant number of iterations ago)
